packages/vue/src/components/organisms/SfHero/SfHero.spec.js

```diff
@@ -0,0 +1,151 @@
+import { shallowMount } from "@vue/test-utils";
+import SfHero from "./SfHero.vue";
+import SfHeroItem from "./_internal/SfHeroItem.vue";
+import SfArrow from "../../atoms/SfArrow/SfArrow.vue";
+import SfBullets from "../../atoms/SfBullets/SfBullets.vue";
+
+const items = [
+  {
+    title: "title_1",
+    subtitle: "subtitle_1",
+    buttonText: "button_1",
+    background: "#eceff1",
+    image: "https://i.ibb.co/6HS24vc/hero-bg-removebg-preview.png"
+  },
+  {
+    title: "title_2",
+    subtitle: "subtitle_2",
+    buttonText: "button_2",
+    background: "#eceff1",
+    image: "https://i.ibb.co/6HS24vc/hero-bg-removebg-preview.png"
+  },
+  {
+    title: "title_3",
+    subtitle: "subtitle_3",
+    buttonText: "button_3",
+    background: "#eceff1",
+    image: "https://i.ibb.co/6HS24vc/hero-bg-removebg-preview.png"
+  }
+];
+const singleItem = [
+  {
+    title: "title_single",
+    subtitle: "subtitle_single",
+    buttonText: "button_single",
+    background: "#eceff1",
+    image: "https://i.ibb.co/6HS24vc/hero-bg-removebg-preview.png"
+  }
+];
+
+describe("SfHero.vue", () => {
+  describe("with items array passed as props", () => {
+    let component;
+    beforeEach(() => {
+      component = shallowMount(SfHero, {
+        propsData: {
+          items
+        }
+      });
+    });
+
+    it("renders a component", () => {
+      expect(component.contains(".sf-hero")).toBe(true);
+    });
+
+    it("renders three SfHeroItem components when passed three items", () => {
+      expect(component.findAll(SfHeroItem)).toHaveLength(items.length);
+    });
+
+    it("renders two SfArrow components", () => {
+      expect(component.findAll(SfArrow)).toHaveLength(2);
+    });
+
+    it("renders SfBullets component", () => {
+      expect(component.findAll(SfBullets)).toHaveLength(1);
+    });
+
+    it("calls #go with 'prev' when clicked on left arrow", () => {
+      jest.spyOn(component.vm, "go");
+      expect(component.vm.go).not.toHaveBeenCalled();
+      component
+        .findAll(SfArrow)
+        .at(0)
+        .trigger("click");
+      expect(component.vm.go).toHaveBeenCalledWith("prev");
+    });
+
+    it("calls #go with 'next' when clicked on right arrow", () => {
+      jest.spyOn(component.vm, "go");
+      expect(component.vm.go).not.toHaveBeenCalled();
+      component
+        .findAll(SfArrow)
+        .at(1)
+        .trigger("click");
+      expect(component.vm.go).toHaveBeenCalledWith("next");
+    });
+
+    describe("with 'prev' and 'next' slots populated", () => {
+      let component;
+      beforeEach(() => {
```

Comment:
Hey, I would like to discuss this `beforeEach()` approach. In my opinion, mounting component directly in each test is much more readable, rather than doing it before each test and putting in some mutable variable. I think that because every test should be independent and should contain every necessary information to understand it (maybe expect some bigger input values).

Now when reading just the following test I don't know why you expecting the component to have `forPrevSlot` class.
```
it("renders 'prev' and 'next' slots instead of SfArrow", () => {
        expect(component.findAll(".forPrevSlot")).toHaveLength(1);
        expect(component.findAll(".forNextSlot")).toHaveLength(1);
        expect(component.findAll(SfArrow)).toHaveLength(0);
      });
```
If you would mount component directly in test and introduce input values explicity, then it would be clear why you are expecting component to have this class.
```
 it("renders 'prev' slot instead of SfArrow", () => {
    const testClass = "forPrevSlot"
    const component = shallowMount(SfHero, {
      propsData: { items},
      slots: {
        prev: `<div class='${testClass}'></div>`
      }
    });
    expect(component.findAll(`.${testClass}`)).toHaveLength(1);
    expect(component.findAll(SfArrow)).toHaveLength(0);
  });
```
What do you think about it ?

Reply:
Generally I agree. If this approach saves few lines of code but makes it less readable then it's not a way to go. I'll refactor this.

Reply:
I had a look on it in the code, I understand that it might be less readable for you the way it is right now, but from my perspective code repetition makes it even more unreadable. So I would say it is a personal preference. With using beforeEach, every describe block fits my editor's window, I do not repeat code. Also my describe block message clarifies the tests and their setup included within the describe block. If you want to change configuration of the mounted component you have to do it jut in one place. Those are just few advantages of using beforeEach.
You have also mentioned:
> mounting component (...) and putting in some mutable variable.
>
And here, the way I see it, it is not a problem, that it's stored in mutable variable (and it's an object anyway so even with `const` I can always change the properties). But I should destroy the component after each test, to be sure that I am creating a new one every time in beforeEach method. I have added that to the tests.

Reply:
I see your point of view :smile: And you are right about code repetition in more complex mounting cases.  But as you said it's kind of personal preference. 
For example, I think I would rather export component mounting to some dedicated function, and mount it explicitly in each test, then do it in beforeEach() approach, because as I wrote previously for me it's important to introduce values that you assert in the test, explicitly at the beginning.